app/src/main/java/com/github/HumanLearning2021/HumanLearningApp/model/DataSetInterface.kt

```diff
@@ -1,9 +1,58 @@
 package com.github.HumanLearning2021.HumanLearningApp.model
 
+import android.graphics.drawable.Drawable
+import java.io.Serializable
+
 
 /**
  * An interface representing the part of the model interacting with data sets
  */
 interface DatasetInterface {
+    /**
+     * A function to retrieve a picture from the data set given a category
+     *
+     * @param category the category of the image to be retrieved
+     * @return a CategorizedPicture from the desired category. Null if no picture of the desired
+     * category is present in the dataset.
+     * @throws IllegalArgumentException if the provided category is not present in the dataset
+     */
     suspend fun getPicture(category: Category): CategorizedPicture?
+
+
+    /**
+     * A function that allows to put a picture in the dataset
+     *
+     * @param picture the picture to put in the dataset. Must implement the serializable interface.
+     * A way to do this is by creating a class containing an ID and an array of bytes containing the
+     * PNG data.
+     * @param category the category to which the picture belongs
+     * @return a Categorized picture built using 'picture' and 'category'
+     * @throws IllegalArgumentException if the category provided is not present in the dataset
+     */
+    suspend fun putPicture(picture: Serializable, category: Category): CategorizedPicture
+
+    /**
+     * A function to retrieve a category
+     *
+     * @param categoryName the name of the desired category
+     * @return the desired category if present, null otherwise
+     */
+    suspend fun getCategory(categoryName: String):Category?
+
+    /**
+     * A function to add a category to the dataset. If the provided name matches that of category
```

Comment:
A category's key is not its name in my book, so that constraint is weird to implement for me. It's more natural to create a duplicate category. Obviously the dummy implementation will have problems w/ that though.

Reply:
Not sure I get what you mean. The name here is just the parameter to the function, so components calling the function might want to get a category that corresponds to a particular name. Or do you want them to refer to categories only by their key, which the `putCategory()` function would return?

Reply:
> The name here is just the parameter to the function, so components calling the function might want to get a category that corresponds to a particular name.

There's getCategory for that. They go use that. It does a query. (`SELECT` in SQL parlance). Then they decide if they want to make a new one. `put*` methods are for creating more objects in the database. (`INSERT`)


> Or do you want them to refer to categories only by their key, which the putCategory() function would return?

Ideally yes, otherwise, how do they even know that the given category exists? And what if there's multiple? That's the point of keys.

Reply:
> > The name here is just the parameter to the function, so components calling the function might want to get a category that corresponds to a particular name.
> 
> There's getCategory for that. They go use that. It does a query. (`SELECT` in SQL parlance). Then they decide if they want to make a new one. `put*` methods are for creating more objects in the database. (`INSERT`)

Aren't we talking about `getCategory()`?

Reply:
which would correspong to a `SELECT category where category.name = categoryName` roughly no?

Reply:
Yes precisely. Now what I'm complaining about is that currently, putCategory is `SELECT`, then conditionally `INSERT INTO category` based on the wrong assumption that category names should be unique database-wide.

Reply:
Ah okay misunderstanding then, I didn't get that you were talking about `putCategory`

Reply:
> Yes precisely. Now what I'm complaining about is that currently, putCategory is `SELECT`, then conditionally `INSERT INTO category` based on the wrong assumption that category names should be unique database-wide.

But then wouldn't it be possible to create a duplicate category with the same parameters (for now name, but we could imagine more), insert it unconditionally into the database with a different key, and return a reference to another category object to the caller?